stacks-signer/src/ping/periodic_pinger.rs

```diff
@@ -0,0 +1,102 @@
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        mpsc::Sender,
+        Arc,
+    },
+    thread::{self, JoinHandle},
+    time::Duration,
+};
+
+use crate::runloop::RunLoopCommand;
+use tracing::{info, trace, Span};
+
+/// Stop the pinger thread
+pub struct ThreadStopHandle {
+    keep_running: Arc<AtomicBool>,
+    handle: JoinHandle<()>,
+}
+
+/// Send a ping command periodically
+pub struct PeriodicPinger {
+    keep_running: Arc<AtomicBool>,
+    wait_for: Duration,
+    sender: Sender<RunLoopCommand>,
+    payload_size: u32,
+}
+
+impl PeriodicPinger {
+    /// Spawn a new pinger thread
+    pub fn spawn(
+        tx: Sender<RunLoopCommand>,
+        wait_for: Duration,
+        payload_size: u32,
+    ) -> ThreadStopHandle {
+        let runner = Self {
+            keep_running: Arc::new(AtomicBool::new(true)),
+            wait_for,
+            sender: tx,
+            payload_size,
+        };
+
+        let span = Span::current();
+
+        ThreadStopHandle {
+            keep_running: Arc::clone(&runner.keep_running),
+            handle: thread::Builder::new()
+                .name("periodic-pinger".into())
+                .spawn(move || {
+                    let _current = span.entered();
+                    runner.run()
+                })
+                .unwrap(),
+        }
+    }
+
+    fn run(self) {
+        while self.keep_running() {
+            if let Err(e) = self.sender.send(RunLoopCommand::Ping {
+                payload_size: self.payload_size,
+            }) {
+                info!("Exit: Send channel closed: {e}");
+                break;
+            }
+            trace!("Sent ping cmd");
+            thread::sleep(self.wait_for);
+        }
+    }
+
+    fn keep_running(&self) -> bool {
+        self.keep_running.load(Ordering::Relaxed)
+    }
+}
+
+impl ThreadStopHandle {
+    /// Stop the open loop and join the thread handle
+    pub fn stop(self) {
+        self.keep_running.store(false, Ordering::Relaxed);
```

Comment:
Why `Ordering::Relaxed` and not `Ordering::SeqCst`?

Reply:
In my experience, simple toggles don't require extra consistency guarantees. I'll change it anyway.